framework/src/loops/ComputeJacobianThread.C

```diff
@@ -64,14 +62,6 @@ ComputeJacobianThread::computeJacobian()
       {
         kernel->prepareShapes(kernel->variable().number());
         kernel->computeJacobian();
-        /// done only when nonlocal kernels exist in the system
-        if (_fe_problem.checkNonlocalCouplingRequirement())
-        {
-          std::shared_ptr<NonlocalKernel> nonlocal_kernel =
-              std::dynamic_pointer_cast<NonlocalKernel>(kernel);
-          if (nonlocal_kernel)
-            kernel->computeNonlocalJacobian();
```

Comment:
We don't need this? How would the nonlocal jacobians be computed then?

Reply:
In all existing uses of nonlocal kernels, the (preconditioning) matrix is something other than diagonal-only. `ComputeJacobianThread` is only used for diagonal-only matrices. `ComputeFullJacobianThread` is used in all other cases, so for instance if you add a `SingleMatrixPreconditioner` then you will get a `ComputeFullJacobianThread`, which still calls `kernel->computeNonlocalJacobian()`

Reply:
I see. So, maybe we should at least add a warning here, so that in the future if someone tries to use nonlocal kernels with a diagonal-only preconditioner is notified about this.

Reply:
I can add an error